src/charm.py

```diff
@@ -1054,59 +1054,45 @@ def _configure_pki_secrets_engine(self) -> None:
-            self._set_pki_csr_secret(csr)
-
-    def _is_intermediate_ca_common_name_valid(self, vault: Vault, common_name: str) -> bool:
-        """Check if the intermediate CA is set with the valid common name."""
-        intermediate_ca = vault.get_intermediate_ca(mount=VAULT_PKI_MOUNT)
-        if not intermediate_ca:
-            return False
-        intermediate_ca_common_name = get_common_name_from_certificate(intermediate_ca)
-        return intermediate_ca_common_name == common_name
-
-    def _is_intermediate_ca_set(self, vault: Vault, certificate: str) -> bool:
-        """Check if the intermediate CA is set in the PKI secrets engine."""
-        intermediate_ca = vault.get_intermediate_ca(mount=VAULT_PKI_MOUNT)
-        return certificate == intermediate_ca
-
-    def _add_intermediate_ca_certificate_to_pki_secrets_engine(self) -> None:
-        """Add the CA certificate to the PKI secrets engine."""
-        if not self.unit.is_leader():
-            logger.debug("Only leader unit can handle a vault-pki certificate request")
+        config_common_name = self._get_config_common_name()
+        if not config_common_name:
+            logger.error("Common name is not set in the charm config")
             return
-        vault = self._get_active_vault_client()
-        if not vault:
-            logger.debug("Vault is not ready to handle a vault-pki certificate request")
+        certificate_request = self._get_certificate_request()
+        if not certificate_request:
+            logger.error("Certificate request is not valid")
             return
-        certificate = self._get_pki_intermediate_ca_certificate()
-        if not certificate:
+        provider_certificate, private_key = self.tls_certificates_pki.get_assigned_certificate(
+            certificate_request=certificate_request
+        )
+        if not provider_certificate:
             logger.debug("No certificate available")
             return
-        common_name = self._get_config_common_name()
-        if not common_name:
-            logger.error("Common name is not set in the charm config")
+        if not private_key:
+            logger.debug("No private key available")
             return
-        if not self._is_intermediate_ca_common_name_valid(
-            vault, common_name
-        ) or not self._is_intermediate_ca_set(vault, certificate):
-            vault.set_pki_intermediate_ca_certificate(
-                certificate=certificate, mount=VAULT_PKI_MOUNT
-            )
+        vault.enable_secrets_engine(SecretsBackend.PKI, VAULT_PKI_MOUNT)
+        existing_ca_certificate = vault.get_intermediate_ca(mount=VAULT_PKI_MOUNT)
+        if (
+            existing_ca_certificate
+            and Certificate.from_string(existing_ca_certificate)
+            == provider_certificate.certificate
+        ):
+            logger.debug("CA certificate already set in the PKI secrets engine")
+            return
+        self.vault_pki.revoke_all_certificates()
```

Comment:
I'm not sure about this. We don't have a clear policy on handling CA changes in Vault and I don't see why is revoking all certificates when the CA changes better than waiting them to expire.
I'm not opposing it, but I'd prefer this to be part of a well defined policy.

Reply:
[TE098](https://docs.google.com/document/d/1zSpr7H7Oyttma73J8IEgdfEiGgzdnQdZX79evzGfHec/edit) suggests revoking certificates if the CA is revoked or invalidated. But we aren't doing that here to my understanding, the "old" CA will still be there.

Reply:
We were already doing so in the vault k8s charm (prior to using the tls v4 lib), here we simply bring the two to parity. I'm not necessarily against changing the behavior but I'd scope this out of this task.

Reply:
Alright.

Reply:
Ticket created.